config/src/config/network_config.rs

```diff
@@ -340,58 +340,95 @@ impl Default for RateLimitConfig {
+#[derive(Clone, Copy, Debug, Deserialize, PartialEq, Serialize)]
+pub enum PeerRole {
+    Validator = 0,
+    PreferredUpstream,
+    ValidatorFullNode,
+    Upstream,
+    Unknown,
+}
+
+impl Default for PeerRole {
+    /// Default to least trusted
+    fn default() -> Self {
+        PeerRole::Unknown
+    }
+}
 
 /// Represents a single seed configuration for a seed peer
 #[derive(Clone, Debug, Default, Deserialize, PartialEq, Serialize)]
 #[serde(default)]
-pub struct TrustedPeer {
-    pub keys: HashSet<x25519::PublicKey>,
+pub struct Peer {
     pub addresses: Vec<NetworkAddress>,
+    pub keys: HashSet<x25519::PublicKey>,
+    pub role: PeerRole,
 }
 
-impl TrustedPeer {
+impl Peer {
     /// Combines `Vec<NetworkAddress>` keys with the `HashSet` given
     pub fn new(
         addresses: Vec<NetworkAddress>,
         mut keys: HashSet<x25519::PublicKey>,
-    ) -> TrustedPeer {
+        role: PeerRole,
+    ) -> Peer {
         let addr_keys = addresses
             .iter()
             .filter_map(NetworkAddress::find_noise_proto);
         keys.extend(addr_keys);
-        TrustedPeer { addresses, keys }
+        Peer {
+            addresses,
+            keys,
+            role,
+        }
     }
 
-    /// Combines two `TrustedPeer`.  Note: Does not merge duplicate addresses
-    pub fn extend(&mut self, other: TrustedPeer) {
+    /// Combines two `Peer`.  Note: Does not merge duplicate addresses
+    /// TODO: Instead of rejecting, maybe pick one of the roles?
+    pub fn extend(&mut self, other: Peer) -> Result<(), Error> {
+        crate::config::invariant(
+            self.role != other.role,
+            format!(
+                "Roles don't match self {:?} vs other {:?}",
+                self.role, other.role
+            ),
+        )?;
         self.addresses.extend(other.addresses);
         self.keys.extend(other.keys);
+        Ok(())
     }
-}
 
-impl From<Vec<NetworkAddress>> for TrustedPeer {
-    /// Converts a `Vec<NetworkAddress>` to a `TrustedPeer`s by extracting the `x25519::PublicKey`s from the address
-    fn from(addresses: Vec<NetworkAddress>) -> TrustedPeer {
+    pub fn from_addrs(role: PeerRole, addresses: Vec<NetworkAddress>) -> Peer {
         let keys: HashSet<x25519::PublicKey> = addresses
             .iter()
             .filter_map(NetworkAddress::find_noise_proto)
             .collect();
-        TrustedPeer { addresses, keys }
+        Peer::new(addresses, keys, role)
     }
 }
 
-impl From<NetworkAddress> for TrustedPeer {
-    fn from(address: NetworkAddress) -> TrustedPeer {
-        TrustedPeer::from(vec![address])
-    }
+/// TODO: This is to simplify the context around auth, but couldn't we just pass the original object?
+#[derive(Clone, Debug, Default, Deserialize, PartialEq, Serialize)]
+pub struct PeerAuthContext {
+    pub keys: HashSet<x25519::PublicKey>,
+    pub role: PeerRole,
```

Comment:
there's probably not a good enough reason to keep these separate anymore. I would be fine if this was just `Peer`